### ISSUE TITLE
Proposal: Remove fixed Kotlin runtime version limitation in Gradle

### ISSUE THREAD

**Issue by ZacSweers** [gradle/gradle#16345]
Today Gradle bundles an embedded version of Kotlin. My understanding is that this is due to usage of private compiler APIs in order to support kotlin-dsl. This prevents _all_ plugins on the classpath from using different (let alone newer) Kotlin versions though, even if both they and the consuming build are not using kotlin-dsl.

This is something that I feel is fundamentally holding back the ecosystem, because it negatively affects plugin developers in the following ways:
* They _must_ set arbitrary language settings on their build, like lowering the `languageLevel` setting to `1.3` today, otherwise risk accidentally writing incompatible code.
* For much of Kotlin 1.4.x's release cycle, they could not use Kotlin 1.4 period. Kotlin 1.4 was not added until Gradle 6.8, and even then it didn't allow kotlin 1.4 language targeting.
* Language level alone only covers their own sources, but then they must _also_ meticulously test any Kotlin-using external libraries that may have updated to 1.4 and use 1.4 APIs.
  * A good example of where this comes up is Okio, OkHttp, or Coroutines. Even something as simple as a member reference becomes a breaking change when compiled in 1.4 and run in a 1.3 environment.

This would be less of an issue if Gradle updated its embedded Kotlin version regularly, but the judgement today is that it cannot target a new minor version of Kotlin (now on a ~6mo cadence) until a new _major_ version of Gradle (~once per year). While this may have lined up somewhat with Kotlin's previous ~annual release strategy, this is unwinding now with Kotlin 1.5's upcoming release. As of now, Gradle is on track to miss the entirety of Kotlin 1.4 due to its reluctance to update or allow users to use a new version. This will only continue now that Kotlin's release cycle is accelerated.

I understand that part of this reluctance to allow newer versions is because they may break. However, I think this is easily remedied by simply making it clear what the currently supported version is. Gradle adopts this approach with Java today, where it often supports _running_ on a newer version but may not _officially_ support it until a later release (major or otherwise). [Gradle added Java 15 support in the 6.7 minor release, for instance](https://docs.gradle.org/6.7-rc-1/release-notes.html).

The private API usage does offer a good reason for using a fixed version, but again there is precedent for this too as Gradle uses private javac APIs too (namely from the private-ish tools jar) to support cases like incremental compilation.

I would like to propose adopting one of the following strategies (in order of preference, though they're not all mutually exclusive either):

1. Unbundle Kotlin, remove private API usage or move it behind a flag. "If you want to use these APIs in kotlin-dsl, this must be enabled and comes with these restrictions"
2. Unbundle Kotlin, use a simple minimum/supported version and denote new un-tested versions as unsupported "your mileage may vary"
3. Allow/commit to more frequent updates. I.e. allow minor Kotlin version updates in minor Gradle version updates
4. Unbundle Kotlin, shade the embedded kotlin compiler or its privately used APIs to avoid conflict

**Comment by CristianGM** [comment]
I would say my preference order is:
3
2
1
4

**Comment by bamboo** [comment]
> This prevents all plugins on the classpath from using different (let alone newer) Kotlin versions though, even if both they and the consuming build are not using kotlin-dsl.

The statement above is not true. 

Currently Gradle only exposes (or should only expose) the Kotlin runtime (basically `kotlin-stdlib*` and `kotlin-reflect*`, no compiler libs) to the plugin/script classpath. So while plugins that, for one reason or another, are incompatible with the particular version of the Kotlin stdlibs shipped with Gradle are not supported, this is not the same as preventing all plugins on the classpath from using different Kotlin versions.  

That being said, I do appreciate how inconvenient and far from ideal the status quo is and agree we should find a better strategy.

**Comment by ZacSweers** [comment]
Fair! In practice though - restricting both the target language version and stdlib (kotlin's "runtime") is effectively the same as limiting the "language version" for Kotlin in my opinion, just a semantic difference. Adding kotlin-reflect and other core libraries adds to that as well as people expect them to match the core "Kotlin" version (i.e. 1.4.30)

**Comment by ZacSweers** [comment]
Another example of where this causes friction: in most repos, it ends up requiring maintaining brittle infrastructure to support two kotlin versions. If a repo chooses to move ahead, it faces noisy, obscure warnings from kotlinc like this:

```
w: Runtime JAR files in the classpath should have the same version. These files were found in the classpath:
    /home/runner/.gradle/wrapper/dists/gradle-7.0-all/9m115ut5nwvtxli7nys8pggfr/gradle-7.0/lib/kotlin-stdlib-1.4.31.jar (version 1.4)
    /home/runner/.gradle/wrapper/dists/gradle-7.0-all/9m115ut5nwvtxli7nys8pggfr/gradle-7.0/lib/kotlin-stdlib-common-1.4.31.jar (version 1.4)
    /home/runner/.gradle/wrapper/dists/gradle-7.0-all/9m115ut5nwvtxli7nys8pggfr/gradle-7.0/lib/kotlin-stdlib-jdk7-1.4.31.jar (version 1.4)
    /home/runner/.gradle/wrapper/dists/gradle-7.0-all/9m115ut5nwvtxli7nys8pggfr/gradle-7.0/lib/kotlin-stdlib-jdk8-1.4.31.jar (version 1.4)
    /home/runner/.gradle/wrapper/dists/gradle-7.0-all/9m115ut5nwvtxli7nys8pggfr/gradle-7.0/lib/kotlin-reflect-1.4.31.jar (version 1.4)
    /home/runner/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk8/1.4.31/e613be5465ef1e6fd0468707690b7ebf625ea2fe/kotlin-stdlib-jdk8-1.4.31.jar (version 1.4)
    /home/runner/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-reflect/1.4.31/63db9d66c3d20f7b8f66196e7ba86969daae8b8a/kotlin-reflect-1.4.31.jar (version 1.4)
    /home/runner/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-jdk7/1.4.31/84ce8e85f6e84270b2b501d44e9f0ba6ff64fa71/kotlin-stdlib-jdk7-1.4.31.jar (version 1.4)
    /home/runner/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.5.0-M2/575e1b43819635280e68bd3d3e8441603e97bbc3/kotlin-stdlib-1.5.0-M2.jar (version 1.5)
    /home/runner/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.5.0-M2/5e6ad08bde4c95d93ca6b366ca5e3a0bb49c70bc/kotlin-stdlib-common-1.5.0-M2.jar (version 1.5)
w: Consider providing an explicit dependency on kotlin-reflect 1.5 to prevent strange errors
w: Some runtime JAR files in the classpath have an incompatible version. Consider removing them from the classpath
e: warnings found and -Werror specified

> Task :agp-handlers:agp-handler-42:kaptGenerateStubsKotlin

FAILURE: Build failed with an exception.
```

Even worse - it's per-project _and_ it prevents use of `-Werror`.

**Comment by Tapchicoma** [comment]
@ZacSweers this is the problem on KGP side: https://youtrack.jetbrains.com/issue/KT-41142

**Comment by ZacSweers** [comment]
The [Forward Compatibility](https://docs.gradle.org/current/userguide/kotlin_dsl.html#forward_compatibility) section of the kotlin-dsl docs mention this:

> The biggest issue is the compatibility between the external kotlin-gradle-plugin version and the kotlin-stdlib version shipped with Gradle. More generally, between any plugin that transitively depends on kotlin-stdlib and its version shipped with Gradle. As long as the combination is compatible everything should work. This will become less of an issue as the language matures.

Kotlin as a language is mature today and guarantees backward compatibility. Is Gradle claiming to add an extra layer of compatibility or is this doc out of date?

**Comment by eygraber** [comment]
I've been tearing my hair out trying to figure this out. Thanks @ZacSweers for combining all of this info in one place (too bad I found it after gathering all the info myself).

I've been trying to update Detekt to Kotlin 1.5.0 (https://github.com/detekt/detekt/pull/3718), and this has been causing lots of problems because of `-Werror`. 

I've also spent the past few days updating one of my projects to use gradle kotlin dsl, and this has been a pretty big pain point (even just figuring out what's going on).

**Comment by yyjdelete** [comment]
For the warn `Runtime JAR files in the classpath should have the same version`
Seems it's also reported as https://youtrack.jetbrains.com/issue/KT-41142 and already be fixed in `org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.10`

**Comment by eygraber** [comment]
I thought that as well, but seems like it's a different issue (related to the Kotlin Gradle plugin's dependency).

I upgraded to Kotlin 1.5.10 and this issue still occurs for me.

**Comment by Tapchicoma** [comment]
@eygraber please open new Kotlin issue for your case

**Comment by eygraber** [comment]
@Tapchicoma my issue is that Gradle is still using Kotlin 1.4.31 but I'm using Kotlin 1.5.X. 

My understanding is that's what this issue is about.

**Comment by eygraber** [comment]
I've investigated further, and found that my specific issue comes from the `java-gradle-plugin` (specifically that it adds `gradleApi()` as a dependency).

So it seems that if you are building a Gradle plugin you need to have that project specifically use whatever version of Kotlin is being used by your version of Gradle.

That's not ideal because it means that a plugin that is being developed as part of bigger ecosystem can't share common build logic with the rest of the projects. If that's unavoidable then there should be better documentation/warnings around this.

**Comment by LouisCAD** [comment]
And the Gradle init template should also reflect that.

**Comment by eygraber** [comment]
Just to add on to my previous comment, not only can the plugin project not share common build logic with the rest of the projects, but none of the other projects can use buildSrc to configure Kotlin since adding the implementation dependency in buildSrc will cause this issue with the plugin project (see https://github.com/detekt/detekt/pull/3835 as a consequence of this).

The only way I can think of to achieve this would be to make the plugin an entirely separate gradle project and include the other project as an included build (although I'm not sure in that case if the incompatible version of Kotlin would be included in the classpath).

**Comment by asarkar** [comment]
As a plugin author, I support @ZacSweers on this ticket. I don't want to be told by Gradle what Kotlin version to use. My order of preference is 1, 3, 4, 2.

**Comment by martinbonnin** [comment]
@eygraber FWIW, you can compile a Gradle plugin with Kotlin 1.5 without the warning by using `dev.gradleplugins:gradle-api:7.1.1` instead of `gradleApi()` in your dependencies. Something like:

```kotlin
dependencies {
  compileOnly("dev.gradleplugins:gradle-api:7.1.1")
}

// remove gradleApi() automatically added by `java-gradle-plugin`
// This is fragile because it's based on the name of the dependency. Maybe it would be easier to not apply
// `java-gradle-plugin` in the first place
val apiDependencies = project.configurations.getByName("api").dependencies
val gradleApi = apiDependencies.firstOrNull {
  val targetComponentId = (it as? DefaultSelfResolvingDependency)?.targetComponentId?.toString()
  targetComponentId == "Gradle API"
}

if (gradleApi != null) {
  apiDependencies.remove(gradleApi)
}
```

`dev.gradleplugins:gradle-api:7.1.1` artifacts are [redistributed by Nokee](https://docs.nokee.dev/manual/gradle-plugin-development.html#sec:gradle-dev-redistributed-gradle-api) (see also [this issue](https://github.com/gradle/gradle/issues/1835)) and are resolved together with all other dependencies. Meaning if two versions of `kotlin-stdlib` are in the dependencies, they will be resolved according to [Gradle conflict resolution rules](https://docs.gradle.org/current/userguide/dependency_resolution.html#sub:resolution-strategy) and the highest one will be selected (1.5 here). I'm not sure why the warning happens with `gradleApi()` but it looks like it somehow bypasses the standard conflict resolution rules.

**Comment by ZacSweers** [comment]
The higher one won't be selected because Gradle always forces the version it uses internally

**Comment by martinbonnin** [comment]
@ZacSweers it won't change anything at runtime (when executing the build) but I'm pretty sure it will remove the warning when compiling the plugin. 

A plugin build is actually pretty similar to any other Kotlin lib build. I found out the only reason it output warning was because `java-gradle-plugin` was forcing `api(gradleApi())`. Not 100% what's that doing behind the scene but it seems like it forces another `kotlin-stdlib` version in the compile classpath.

**Comment by pioterj** [comment]
@melix You posted a great explanation of this issue and why it currently works this way on the Android Study Group Slack. Would you mind pasting it here?

**Comment by melix** [comment]
_I cannot speak for the Gradle team anymore, but this is the answer I gave to this question in a private group. It's adapted from a conversation so please bare with me with the layout. I will let the Gradle team answer/comment on the rest :)_

You cannot change the version of Kotlin that Gradle uses. It's a "runtime": Kotlin build scripts have to be compiled, so Gradle has to come with one version of Kotlin. Your plugins, if written in Kotlin, therefore have to use the same version of Kotlin, because that's the runtime they will use for their own classes: the build scripts will be compiled by Gradle using the embedded version, and the same version is used to load your plugin. This doesn't mean that your plugins can't use a different Kotlin version for their task actions: use the worker API to isolate things properly. And it doesn't mean either that your apps can't use a different Kotlin version: because the compiler is executed in a different daemon, that's not a problem. In other words, the only issue is for your own plugin classes.

And to be clear, when I mean to load your plugin, I mean think about code like `tasks.withType<MyTaskWrittenInKotlin>`: the type `MyTaskWrittenInKotlin` is using Kotlin internal types which are also used by the `withType` extension method: they are effectively sharing the same Kotlin runtime. What happens within the task action is under your control, though. There you can isolate things properly using the worker API.

When you write:

```
plugins {
    id("my-plugin-written-in-kotlin")
}
```

the plugins block is compiled... in Kotlin. So Gradle has to come with a runtime with Kotlin. That's the embedded version. This script will resolve the plugins block, but as you are aware there are many others (settings, buildscript block, ...). All in all, it's during that phase that the plugins are resolved and that their classpath is computed. As a complicating factor, plugins are cached by classloader, by set of plugins applied, so different sets of plugins on different subprojects may have different resolved versions.
All in all, it means that even before we know that we need to apply your plugin, Kotlin is already initialized and evaluated. Then, the full build script needs to be compiled. And there, depending on what you have on classpath, you may have plugins written in different versions of Kotlin which require different levels of language. This is no surprise: Gradle for example had to enable experimental flags to support script compilation, or enable certain type of inference, ... Meaning that the horrible truth is that different Kotlin versions are not compatible.

To some extent, this is very similar to this problem: take a plugin written in Java, compiled for Java 16. Try to run a Gradle build with Java 8: it will blow up, because class format is not compatible. There's simply nothing we can do against this, except telling plugin authors: try to use the lowest version of Java possible to maximize compatibility.

Because you can't switch the Java version of the Gradle runtime: it's simply not possible: the user has to run on a higher version in this case. So why I understand the frustration, there are good reasons why it's designed this way, and no, it's not easy to change. Groovy users have lived for years with this limitation tbh (for plugins written in Groovy) but this is also why, in my personal case, am trying to write my plugins in pure Java now, to avoid those problems. And btw, I'm not even talking about the cost of having multiple language runtimes at execution time: I'd bet you would end up complaining that Gradle uses too much resources.

So, the rules are simple:

- you want to write your plugin in Kotlin? Fine, live with the restriction: use the embedded version
- you want to support executing task actions with any version of Kotlin? use the worker API

We cannot have our cake and eat it too...

> Could Gradle compile the script using 1.4 in a first step. And during execution use 1.5 if some plugin pulls it in its classpath? Code compiled with 1.4 compiler should run fine with stdlib 1.5? 

I guess in theory it's doable. But, what a nightmare. As I said the compiler options that Gradle has to inject are different for different Kotlin versions, because features are experimental in one release, not in another, etc. The script compilation infrastructure is already quite complex, so if it has to know about different Kotlin versions and the user classpath in addition, that's becoming too complex. That's just an incentive for having weird compilation errors and very hard to diagnose bugs. And I'm not even mentioning performance. Kotlin compilation is slow enough, I'm not sure that spawning even more compilation daemons just for script compilation is a good idea. But, I live that decision to the Gradle team.

**Comment by emartynov** [comment]
Gradle just runs on the specified JDK. Why not allow something similar for the Kotlin? 

It will give some freedom for the consumer to use, play, fix report things that are not compatible. He/she can do something proactive.

However, that would open the box of the Pandore - why not do similar for Groovy (I don't know what other runtime dependencies Gradle needs).

**Comment by LouisCAD** [comment]
What we want is the ability to specify the Kotlin stdlib version in `gradle.properties` or in a shebang in `settings.gradle[.kts]`.

That's all we need to unblock use of libraries that follow the Kotlin train from within Gradle plugins that are not using isolated workers.

That wouldn't require to upgrade dependencies on the compiler within Gradle since newer stdlibs are a superset of the older ones.

**Comment by melix** [comment]
> What we want is the ability to specify the Kotlin stdlib version in gradle.properties or in a shebang in settings.gradle[.kts]

It breaks the principle of isolation: a user shouldn't have to know what version of Kotlin to use to build a project. It may be that a transitive dependency uses a higher version and you wouldn't notice. This isn't, IMHO, a viable solution.

**Comment by ZacSweers** [comment]
We do that today with java versions but Gradle doesn't impose a specific version. Many plugins use guava and users accept that they may need to address transitive version mismatches. Yet, kotlin must abide by different rules because Gradle uses it internally. I think that's pretty unfortunate.

A good example of this is happening right now with Gradle 7.2 - it updates the embedded kotlin version to 1.5.21 but doesn't offer guidance to plugin authors. Should they continue to set their language/API version to 1.4? Can they raise it only if their min Gradle version is 7.2+? Etc etc.

I'd understand if it was a default but the fact that Gradle actively prevents changing the version is problematic to the community at large because it results in fragmented versioning and frankly most plugin authors don't realize this is even going to be a problem until they're burned by it.

**Comment by LouisCAD** [comment]
> > What we want is the ability to specify the Kotlin stdlib version in gradle.properties or in a shebang in settings.gradle[.kts]
> 
> It breaks the principle of isolation: a user shouldn't have to know what version of Kotlin to use to build a project. It may be that a transitive dependency uses a higher version and you wouldn't notice. This isn't, IMHO, a viable solution.

I'm going to use these words against that statement itself:
Kotlin libraries and their developers shouldn't have to know what Kotlin version to use just to make sure that they can be used in Gradle plugins.

IMHO, this is a tradeoff, and as a user, I'd prefer to give up that "principle of isolation" when opting-in to the latest Kotlin stdlib version in my Gradle project, so I can use recent libraries and some Gradle plugins with the latest features (and fixes?).

**Comment by melix** [comment]
> IMHO, this is a tradeoff, and as a user, I'd prefer to give up that "principle of isolation" when opting-in to the latest Kotlin stdlib version in my Gradle project, so I can use recent libraries and some Gradle plugins with the latest features (and fixes?).

I heartfully disagree with this: as a user, I want to apply a plugin and I don't care what language it's written in. Forcing me to put a Kotlin version so that the plugin works is counter-intuitive. As a plugin author, *you* care, but users shouldn't.

**Comment by dimsuz** [comment]
Some users (or many?) write mini-plugins in buildSrc, in kotlin. So they still end up having to care.

Also I guess those of us who are also using gradle to build Kotlin projects, suffer from this issue more, because you have to work with different kotlin versions in one project. If I'm not mistaken most of us here who are in favor of this issue are exactly from this category of users.

**Comment by Mahoney** [comment]
Yes, there are two distinct use cases here:
* writing plugins for general purpose reuse
* writing plugins only for use in the current build

In the latter case it _might_  make sense to be able to tell Gradle what version of Kotlin to include so that you can write your plugins in the same version as the actual source.

Worth noticing, however, that you'd be quite likely to run into issues with 3rd party plugins you use that were written in Kotlin and depend on features that Kotlin has removed in a later version... Mind you, this presumably becomes an issue when Gradle upgrades its Kotlin version. Non-backwards compatibility is a big deal. The Java stdlib gets that more right than most.

**Comment by ZacSweers** [comment]
> depend on features that Kotlin has removed in a later version

This, again, exists with any dependency or major java release. 

It's also a frankly nonsensical argument to use kotlin as an example of that as it guarantees backward compatibility across its minor versions and there's no inkling of a Kotlin 2.0 coming any time soon. Let's keep to realistic examples.

**Comment by renatoathaydes** [comment]
We have some of our own plugins written in Kotlin and as we upgrade Gradle, things tend to break because we enforce a policy of no version conflict in the classpath.

We end up doing this in all Kotlin projects:

```groovy
// this is necessary to fix a conflict in the Kotlin stdlib dependencies:
// -+- org.jetbrains.kotlin:kotlin-stdlib:1.5.10
//  \-- org.jetbrains.kotlinx:kotlinx-coroutines-core:1.3.8 (requested org.jetbrains.kotlin:kotlin-stdlib:1.3.71)
configurations {
    kotlinCompilerPluginClasspathMain {
        resolutionStrategy.force "org.jetbrains.kotlin:kotlin-stdlib:$kotlinCompilerVersion"
        resolutionStrategy.force "org.jetbrains.kotlin:kotlin-stdlib-common:$kotlinCompilerVersion"
    }
    kotlinCompilerPluginClasspathTest {
        resolutionStrategy.force "org.jetbrains.kotlin:kotlin-stdlib:$kotlinCompilerVersion"
        resolutionStrategy.force "org.jetbrains.kotlin:kotlin-stdlib-common:$kotlinCompilerVersion"
    }
}
```

It's ugly but fixes the conflicts.

**Comment by LouisCAD** [comment]
@renatoathaydes Looks like you do this in a Groovy script. settings.gradle, build.gradle, or an init script?

That gets me thinking that an init script could do the job of forcing a recent Kotlin stdlib to ensure compatibility with libraries. It could even be exported as a plugin (`Plugin<Gradle>`).

**Comment by ZacSweers** [comment]
@renatoathaydes I believe you're talking about something unrelated to this issue. The build itself will still be forced to use the version Gradle embeds, there's no current way to force this resolution to another version (or else this issue wouldn't exist!). You can write a Gradle plugin in whatever language you want, that doesn't change the fact that it's forced to the runtime of whatever Kotlin version the consuming project's Gradle version embeds.

**Comment by renatoathaydes** [comment]
@LouisCAD `build.gradle` for every project using Kotlin.

@ZacSweers I believe it's the same issue because when I add a plugin to the buildScript classpath, it introduces a conflict with what I believe is Gradle's built-in Kotlin (which breaks our build as we use a strict strategy that forces us to resolve versions "manually"  - this may be harder than it needs to be, but it's current policy due to nasty issues we've had before). 

Once we get a conflict, there's two options:

* look at every plugin's transitive dependencies to see which one is causing it, then exclude the problematic transitive dependency.
* force Gradle to use a particular version as I show above.

I actually prefer to do the first option, but for some reason, Gradle does not show me which plugins are actually conflicting, which is why I was forced to use the other option.

I understand that, as you say, at runtime Gradle will use whatever Kotlin it wants, but my problem is when trying to compile my project's build script (as the `kotlinCompilerPluginClasspathMain` configuration resolution is what fails).

Our build is bloody complex though, so I might be missing something.

**Comment by cortinico** [comment]
> is happening right now with Gradle 7.2 - it updates the embedded kotlin version to 1.5.21 but doesn't offer guidance to plugin authors

Seconding this. As a plugin author there is little to no guidance on how to properly setup a Gradle plugin written in Kotlin to offer the higher compatibility to our users.

Would be great if the Gradle team could include the:
- Version of the embedded Kotlin 
- suggested api/language language
In this page for every release: https://docs.gradle.org/7.1.1/userguide/compatibility.html

**Comment by ephemient** [comment]
For what it's worth, this seems to work for me to create an artifact compatible with Gradle's embedded Kotlin version, even in a project using a greater version of the Kotlin plugin:
```kotlin
plugins {
    `embedded-kotlin`
}

dependencies {
    implementation(kotlin("stdlib-jdk8", embeddedKotlinVersion))
}

kotlin.target.compilations.all {
    kotlinOptions {
        apiVersion = embeddedKotlinVersion.split('.').take(2).joinToString(".")
        jvmTarget = JavaVersion.VERSION_8.toString()
    }
}
```

**Comment by Mahoney** [comment]
> > depend on features that Kotlin has removed in a later version
> 
> This, again, exists with any dependency or major java release.

Dependency, yes. But the kotlin stdlib is a rather large dependency.

I don't think I've ever experienced an issue with binary compatibility in the java stdlib in Java 1.3 - 11. Not sure whether it's still as stable now.

> It's also a frankly nonsensical argument to use kotlin as an example of that as it guarantees backward compatibility across its minor versions and there's no inkling of a Kotlin 2.0 coming any time soon. Let's keep to realistic examples.

https://kotlinlang.org/docs/compatibility-guide-15.html#gradually-change-the-return-type-of-collection-min-and-max-functions-to-non-nullable suggests that with the Kotlin 1.6 stdlib binaries a binary compiled against Kotlin 1.4 that relies on `emptyList().min()` returning `null` will start failing with an exception.

**Comment by gildor** [comment]
> that relies on emptyList().min() returning null will start failing with an exception.

It will be binary compatible change, reintroduced API will just have different JVM name

>  I don't think I've ever experienced an issue with binary compatibility in the java stdlib

Is it somehow different in Kotlin? All stable Kotlin versions are binary compatible (excluding experimental APIs)

**Comment by Mahoney** [comment]
>> that relies on emptyList().min() returning null will start failing with an exception.

> It will be a binary compatible change, reintroduced API will just have different JVM name

Cool, if Kotlin are guaranteeing backwards binary compatibility then I withdraw my comment.

Does anyone have a link to an explicit statement of that guarantee? My googling is letting me down.

**Comment by LouisCAD** [comment]
@Mahoney

https://kotlinlang.org/docs/components-stability.html#stability-of-subcomponents

https://kotlinlang.org/docs/language-committee-guidelines.html

**Comment by z0mb1ek** [comment]
I have updated gradle to  7.2-rc-3 and issue has gone

**Comment by ZacSweers** [comment]
@z0mb1ek no, it's not. It probably looks "gone" because the embedded runtime was updated to 1.5.21 and that's what you're using in the project currently. This is just a coincidence and will happen again as soon as the next Kotlin version is out and they are mismatched again.

**Comment by big-guy** [comment]
https://github.com/gradle/gradle/issues/16345 

Hi,

This topic has gotten a lot of attention and discussion. We wanted to share our understanding of the problem, some solutions/challenges and what can be done right now. 

First, let's make sure we're on the same page. When a project "uses Kotlin", this could mean several different things. 

A project may be written in Kotlin.
A project may be written in some other language and uses Kotlin DSL build scripts.
A project may be written in some other language and applies external plugins that are written in Kotlin.
A project may be written in some other language and writes plugins in buildSrc or an included build in Kotlin. 

From Zac's original post, my understanding is that the trouble stems from (3) and (4). Namely, you are trying to write a Gradle plugin in Kotlin or use a library written in Kotlin that uses a version of Kotlin different from the version embedded in Gradle. On top of that, the Kotlin runtime target for Gradle can be lower than what you would like to use. 

For example, Gradle 7.3 embeds Kotlin 1.5.31 with a Kotlin 1.4 target level.  This makes it difficult to write a plugin that wants to use another Kotlin version or a plugin that wants to set a Kotlin 1.5 target level. It also makes it difficult for that plugin to depend on another library that is written in another Kotlin version or targets a different Kotlin target level.

Is that a good summary of the main problem?

Second, what are some of the ways to fix this? 

When we say "Gradle embeds Kotlin", it may be unclear how deep that goes.

Kotlin DSL scripts are written in Kotlin and Gradle uses the Kotlin Gradle Plugin (KGP) and the Kotlin compiler to compile them.
KGP is written in Kotlin.
The integration with KGP is itself written in Kotlin.
Some other parts of Gradle are now written in Kotlin (like the configuration cache).

The dependency on Kotlin is deeper than just the compiler. It's possible to hide some of this from plugins and build scripts, but ultimately, a plugin expects to run with some version of Kotlin available and build scripts are written for a particular version of Kotlin. 

Gradle could be made to isolate every plugin and only expose the version of Kotlin required for each plugin (if needed). Setting aside how this would impact performance, an implication of this approach would be that no plugins could expose Kotlin types as part of their ABI.  This is complicated.

For now, we do not want to make the version of the embedded Kotlin changeable; however, there are other changes we should make to alleviate some of the pain when developing with Kotlin. From the other comments in this issue, we think there are other related problems that could be fixed without changing the embedded Kotlin version.

We should investigate problems around
[Cross-compilation in Kotlin](https://github.com/gradle/gradle/issues/18487): we should provide samples that show how you can use versions of Kotlin that are different from the embedded version for the scenarios described above.
[Multi-version Gradle plugin development in Kotlin](https://github.com/gradle/gradle/issues/18488): we should provide samples that show how you can develop a plugin that targets multiple versions of Gradle (that have different versions of Kotlin). We need to make it clear which version of Kotlin and target level you need to use.
[Composite builds with Kotlin](https://github.com/gradle/gradle/issues/18489): we should provide samples that show how to use buildSrc/composite builds with Kotlin. 

We have good communication with JetBrains, so if there are changes we can push on the KGP side to make this better, we will.

**Comment by CristianGM** [comment]
Perfect summary.

Speaking only about plugins written in kotlin or using dependencies written in kotlin...
Let me ask a naive question, why can't I use a newer version of kotlin in my plugins? Kotlin promises to be backwards compatible in minor releases (the exceptions are incubating/experimental features, but that should be fine), so I expect, for example, to be able to use kotlin 1.6 in my Gradle plugins with Gradle 7.2.

**Comment by big-guy** [comment]
@CristianGM 
> Let me ask a naive question, why can't I use a newer version of kotlin in my plugins? Kotlin promises to be backwards compatible in minor releases (the exceptions are incubating/experimental features, but that should be fine), so I expect, for example, to be able to use kotlin 1.6 in my Gradle plugins with Gradle 7.2.

There are three parts to compatibility. The "target level" I mentioned above, the Kotlin standard library and what version Gradle provides at runtime. We stick with a target level for an entire major release (1.4 for Gradle 7.x), update the version of the standard library in minor releases and expose a single version of Kotlin that's used when executing build scripts. This is the same kind of strategy we use for Groovy, too.

I believe the Kotlin team tries to make it so Kotlin 1.X can target down to 1.(X-2). So you could compile against Kotlin 1.6, target Kotlin 1.4 and run on Gradle, sometimes. If you had a dependency that needed a higher Kotlin target level to work, you may not be able to run on Gradle. What would break would depend on how the embedded Kotlin version would handle the mismatch. 

The other part is the standard library itself. Depending on which versions of Gradle you want to support, you need to constrain your plugin to the API of the lowest version. If you compile against Kotlin 1.6, target Kotlin 1.4 and are careful to only use Kotlin 1.4 APIs, I think this could work. But it's probably easier to compile against Kotlin 1.4.  You _could_ use newer APIs, but you would have to be careful to only use the newer methods in versions of Gradle that had them. 

The last part is the crux of what OP is asking to change. Gradle embeds a particular version of Kotlin and uses that version for the plugins/build logic. If you built your plugin with Kotlin 1.6 and Gradle had a way to swap that version in, it would break other plugins that were written for a different version of Kotlin or Gradle itself (since part of Gradle is written in Kotlin).

**Comment by CristianGM** [comment]
I think what confuses me is:
> If you built your plugin with Kotlin 1.6 and Gradle had a way to swap that version in, it would break other plugins that were written for a different version of Kotlin or Gradle itself (since part of Gradle is written in Kotlin).

Why? If your code worked with 1.4/1.5 it should work fine with 1.6

In any case, I should try to understand the issue better, for sure is more complex than I think.

**Comment by big-guy** [comment]
> I think what confuses me is:
> 
> > If you built your plugin with Kotlin 1.6 and Gradle had a way to swap that version in, it would break other plugins that were written for a different version of Kotlin or Gradle itself (since part of Gradle is written in Kotlin).
> 
> Why? If your code worked with 1.4/1.5 it should work fine with 1.6
> 
> In any case, I should try to understand the issue better, for sure is more complex than I think.

By the time we know that there may be another Kotlin version, we've already started using Kotlin. We would need to isolate the "already running" Kotlin from everything else and this would also need to apply to plugins so that the different versions of Kotlin are not mixed in Gradle's runtime or into other plugins.

**Comment by yogurtearl** [comment]
If we can't specify the kotlin version used in plugins/dsl, next best thing would be to make sure every new gradle release (every 6 to 9 weeks) is updated to use the latest version of kotlin, that will at least reduce the pain a bit. 

Seems Gradle has been doing a good job of keeping up with kotlin releases recently, would love to see kotlin 1.6 bundled with Gradle in the next Gradle version following the 1.6 kotlin release.

**Comment by martinbonnin** [comment]
> Gradle could be made to isolate every plugin and only expose the version of Kotlin required for each plugin (if needed). 

@big-guy coud we have something like this:

- Plugin authors have an easy way to shadow the Kotlin stdlib
- Plugin authors have an easy way to distribute 2 variants of their plugin: one with the "embedded" stdlib and the other packaged as "external"
- When running a build, Gradle selects the variant based on the Plugin variant metadata and current Gradle embedded Kotlin version 

This way, plugin authors that want to use newer versions could ship 2 versions of their plugin jar and depending on the version of Gradle that's running it, the appropriate variant will be selected. There could even be a small message saying that "plugin $xyz is using a newer version of Kotlin that is not supported yet, expect more memory usage"

This is somewhat related to [Multi-version Gradle plugin development](https://github.com/gradle/gradle/issues/18488) except that:

- setup will be easier. Plugin authors wouldn't have to think about which version of Gradle uses what version of Kotlin as Gradle would do the selection automagically.
- The plugin codebase can be 100% using the newest Kotlin version.

Plugins with a relocated stdlib wouldn't be able to use Kotlin types in their ABI but I think most of these types (collections mainly) are aliased to the equivalent Java types so that could be transparent to relocation? Two Kotlin types that could be problematic are `kotlin.Unit` and `kotlin.Metadata` but maybe Jetbrains can make it so that they are binary compatible forever and do not require relocation?

Thoughts?

**Comment by ephemient** [comment]
I think Shadow already works fine for building Gradle plugins: https://imperceptiblethoughts.com/shadow/plugins/#special-handling-of-the-java-gradle-plugin-development-plugin
Last time I tried, I think I had some issues with Kotlin reflection but everything else just worked out of the box.

**Comment by martinbonnin** [comment]
@ephemient shadow has issues where it renames "kotlin" Strings which is problematic as it's also the name of the kotlin extension so any plugin doing something like `extensions.named("kotlin")` will most likely fail (see https://github.com/johnrengelman/shadow/issues/232). Maybe there's a way to fine tune this. This is where an "easy" setup would be nice. 

Looking ahead and beyond relocation itself, it's also about easy publishing and most importantly, Gradle selecting the appropriate plugin variant at runtime based on its current Kotlin version and plugin metadata.

**Comment by ephemient** [comment]
Well. I didn't try applying or working with the Kotlin plugin from within a custom plugin, so I didn't run into that. Does sound like a serious blocker though.

IMO having an unshadowed version for newer Gradle isn't that useful, since (with this approach) you can't use Kotlin types in public API regardless. Do you have a reason I'm missing?

There's a number of other Kotlin types that can leak into public API if you're not careful, e.g. kotlin.jvm.internal.Function*/DefaultConstructorMarker, but they probably don't come up much in Gradle plugins.

**Comment by martinbonnin** [comment]
> you can't use Kotlin types in public API regardless.

This is the part I would like to have a better understanding of. Looks like a bunch of kotlin types in the APIs could still work (Lists, Sets, ...) with the exceptions that you mentioned (`Function*`, `DefaultConstructorMarker`, ...). Maybe that can be enforced with some tooling. 

Overall, I'd very much better like:

* Write your plugin using the Kotlin version of your choice with some (TBD) limitations on your plugin public API

Compared to:

* Not being able to use okio in a plugin because it transitively pulls a Kotlin stdlib version not supported by Gradle

My gut feeling is that the very vast majority of plugins APIs would relocate very well and that the few edges cases could be acceptable or easily workaroundable. I'll try on a test plugin see how that goes.

**Comment by martinbonnin** [comment]
> The community affected by this can easily lower the language/api version, the problem is that we shouldn't need to just because Gradle wants to use Kotlin internally.

I think the problem is how we define 'internally'. My understanding is that there are at least two parts to this:

### Compile-time 

Gradle needs to compile the user `build.gradle.kts` files. This is the 100% internal part. The Kotlin compiler version for this can lag one version behind the latest one most of the times. According to [the evolution rules for binary formats](https://kotlinlang.org/docs/kotlin-evolution.html#evolving-the-binary-format), Gradle lagging one version behind the latest Kotlin version should mostly work:

```
Preferably (but we can't guarantee it), the binary format is mostly forwards compatible with the next feature 
release, but not later ones (in the cases when new features are not used, e.g. 1.3 can understand most 
binaries from 1.4, but not 1.5).
```

This means that if you want to use `T & Any` [available in preview in 1.6.1-M1](https://github.com/Kotlin/KEEP/pull/269) in your plugin for an example, this will mostly fail at compile time because the current 1.5 compiler will not now what to do with this. Note that this only affects the public API of a plugin. If you want to use `T & Any` as an implementation detail, I would expect this to work. As I said [above](https://github.com/gradle/gradle/issues/16345#issuecomment-934127348), I think it's ok to have such limitations in the plugin public API. It's a much lighter restriction than having to write the whole plugin in an old version of Kotlin.

### Run-time

Once Gradle has compiled all the `build.gradle.kts` scripts to some .jar files, it needs to run them. Because these .jar files use Kotlin, they need one version of `kotlin-stdlib.jar` in the classpath. 

This is where the chicken and egg problem kicks in. These .jar files also contain some code that resolves dependencies and might end up needing a more recent version of `kotlin-stdlib` but in order to run the dependency resolution code, they require `kotlin-stdlib`. As @big-guy mentioned [in this comment](https://github.com/gradle/gradle/issues/16345#issuecomment-931707568):

```
By the time we know that there may be another Kotlin version, we've already started using Kotlin. We would need 
to isolate the "already running" Kotlin from everything else and this would also need to apply to plugins so 
that the different versions of Kotlin are not mixed in Gradle's runtime or into other plugins.
```

I'd say this usage isn't really 100% internal anymore because this Kotlin version is the Kotlin version that plugins will see during configuration so plugins are indeed affected by the Kotlin version used by Gradle. 

All in all, it's not one but 2 versions of Kotlin that Gradle uses, the compile-time and run-time. The compile-time is mostly internal. The run-time is the problematic one IIUC. It just happens that they are the same today but this doesn't have to be always the case. Gradle could make an update just after every Kotlin release to bump the run-time version. That would be a path forward but would still put a fixed (albeit more recent) runtime version in the classpath.

**Comment by ZacSweers** [comment]
> By the time we know that there may be another Kotlin version

Wouldn't this be solvable via some configuration? Gradle or system property, for example. It might be a chicken and egg problem at runtime but we still control the hen house too

**Comment by ZacSweers** [comment]
Hi folks, I am stepping back from this issue. Gradle appears intent on continuing to hamstring the use of Kotlin in plugins to protect their own internal use of it (configuration caching, kotlin-dsl, etc). I am not sure why the above proposals are not actionable, and at this point I am beyond frustrated with the backchannel/subtweet messages from Gradle folks that just amount to a vague "_you just don't understand_". You're right, I don't understand. I'd like to, but it's frustrating to be swatted away as a nuisance for asking, especially as a Gradle employee asked me to write this issue up in the first place.

**Comment by asarkar** [comment]
@ZacSweers I can understand, in my experience, Gradle employees don't drive, they ride on high horses, and are immune to logical reasoning from the community. Case in point: https://github.com/gradle/gradle/issues/18006.

**Comment by ljacomet** [comment]
We would like to remind everyone that the [Code of Conduct](https://gradle.org/conduct/) applies here, so please write comments according to it. Part of the answer from @asarkar is unacceptable and in clear violation of that code. There is also definitely no “intent to hamstring the use of Kotlin” by the Gradle team.

This is a complicated issue. We listed several related problems and potential solutions [in this comment](https://github.com/gradle/gradle/issues/16345#issuecomment-931437640). We would like to address those problems, but we don’t think we can do exactly what was asked in the original description at this time.

**Comment by eygraber** [comment]
> Wouldn't this be solvable via some configuration? Gradle or system property, for example. It might be a chicken and egg problem at runtime but we still control the hen house too

Is there any chance something like this will ever be implemented?

I run into this problem constantly, and my internal Gradle plugins are a mess of managing which versions of dependencies they can use, and which have to wait for the next version of Gradle to get updated (unless the library has moved on to a newer version of Kotlin, in which case I now have to go find the newest version that would work with it).

**Comment by JavierSegoviaCordoba** [comment]
Can this be reopened? Or is there any other issue where this is being tracked?

Is there any place (maybe milestones) to check estimated dates or which future Gradle version will have an specific Kotlin version?

Now that Kotlin 1.7.0 and this issue is hard to be resolved, at least knowing in which version of Gradle or date we will be able to use an specific version of Kotlin should be great.

**Comment by JavierSegoviaCordoba** [comment]
@ephemient that is great, hope it adds future versions of Gradle and not only the released ones.
